### PR TITLE
Updates to JSON-based API for stats

### PR DESCRIPTION
**Pull request by captainsafia**
I think I'll do this specific issue in two chunks. First will be purely the JSON API at `/api/stats` and second will be the visual page at `/stats/`.

Let me know how it looks.

### CODE REVIEW

File: orchestrate.py [hunk 1]
```diff
@@ -69,13 +69,16 @@ def get(self, path=None):
...
     def get(self):
         '''Returns some statistics/metadata about the tmpnb server'''
         response = {
                 'available': len(self.pool.available),
                 'capacity': self.pool.capacity,
-                'version': '0.0.1-dev',
+                'version': '0.1.0',
+                'container_image': self.pool.container_config.image,
+                'docker_version': self.spawner.docker_client._docker_client.version(),
+                'docker_containers': self.spawner.docker_client._docker_client.containers()
```
Comment (rgbkrk) [inline orchestrate.py]: This is going to be a long-blocking call that won't hold up well in quick calls for stats on e.g. try.jupyter.org. As much as I'd *want* to see it, this isn't practical in current Docker and Go versions.

File: orchestrate.py [hunk 2]
```diff
@@ -83,6 +86,10 @@ def get(self):
     def pool(self):
         return self.settings['pool']
 
+    @property
+    def spawner(self):
+        return self.settings['spawner']
```
Comment (rgbkrk) [inline orchestrate.py]: What's this spawner for?

**Comment by rgbkrk** [comment]
All in all, *awesome*! I'm looking forward to this.

File: orchestrate.py (hunk 1 above)
Comment (captainsafia) [inline orchestrate.py]: Ditto. I got too excited thinking about all the cool things I was gonna do with D3 and this to think about performance.

I'll take it out for now, but I'll give it more thought in a future iteration of the stats page perhaps?

Comment (rgbkrk) [inline orchestrate.py]: Yeah that's ok. It's only this call to the number of containers and likely only affects the larger deployments (>500 users/containers in the pool). This number is also likely to be redundant as the number of containers equals the capacity + is tmpnb containerized + is the proxy containerized + any another containers.

I'd keep the capacity and available metrics instead.

File: orchestrate.py [hunk 3]
```diff
@@ -232,7 +233,7 @@ def main():
         (r"/spawn/?(/user/\w+(?:/.*)?)?", SpawnHandler),
         (r"/api/spawn/", APISpawnHandler),
         (r"/(user/\w+)(?:/.*)?", LoadingHandler),
-        (r"/stats", StatsHandler),
```
Comment (rgbkrk) [inline orchestrate.py]: Now that I'm thinking about it, we should probably have both /stats and /api/stats use the APIStatsHandler while also changing the tmpnb redirector.

Then folks can do migrations (ourselves included). After the deployments we know about have migrated (if they want to), then we'll change /stats to the new page you're working on.

Comment (minrk) [inline orchestrate.py]: The new human page can be at `/statistics` or `/info` rather than re-using the URL.

**Comment by captainsafia** [comment]
Ditto — updates coming later tonight.

I'll go ahead and include the visualizations on this PR as well.

File: orchestrate.py [hunk 4]
```diff
@@ -69,13 +69,14 @@ def get(self, path=None):
         self.render("loading.html", path=path)
 
 
-class StatsHandler(BaseHandler):
+class APIStatsHandler(BaseHandler):
     def get(self):
```
Comment (rgbkrk) [inline orchestrate.py]: While you're at it, can you add `self.set_header("Content-Type", 'application/json')` to the APIHandler?

Comment (captainsafia) [inline orchestrate.py]: Yes, sir!

effbd25aa080b73d6ccd1a37487a2850

**Comment by rgbkrk** [comment]
From gitter, [a sample codepen that uses metricsgraphics to display metrics over time](http://codepen.io/rgbkrk/pen/WQbeLp).

**Comment by captainsafia** [comment]
@rgbkrk — This is not 100% what I was aiming for, but I think this should be good enough for an initial stats page.

In the future, I'd like to make the plot area-based to better show the proportionally between used and unused containers.

Not sure if you want to merge this and make an initial release or have me modify the visuals on this PR.

**Comment by rgbkrk** [comment]
I'd like to see capacity in here too, but we can follow on.

> In the future, I'd like to make the plot area-based to better show the proportionally between used and unused containers.

Yeah, that would be nice. Still in a time series view, but area based (as a proportion of capacity).

We can always merge this and follow on right after. This hits exactly what we need for the time being.

**Comment by captainsafia** [comment]
Yep! A wise man once said "Let's merge and iterate" :wink:

**Comment by rgbkrk** [comment]
Looking good!

screenshot 2015-09-03 16 56 15

**Comment by rgbkrk** [comment]
(on my 4 node system :wink:)

**Comment by rgbkrk** [comment]
The more I thought about it, the more I realized that showing the volume relationship is a matter of using capacity and available containers, where the labels are adjusted...

**Comment by captainsafia** [comment]
@rgbkrk — What do you mean by that? The way I imagined the final product looking was a stacked area chart (see awful drawing below).

**Comment by rgbkrk** [comment]
You and I are talking about the same thing, yours has the right terminology though. :smile: We definitely want a stacked area chart.

What I'm trying to say is algebraic in manipulation. :wink: The area between available and capacity *is* the used curve. If we plot capacity and available, used is the space between. It's the labels that change (and axes).

**Comment by captainsafia** [comment]
Yep yep yep. Opening up an issue for this. Thinking it might be in the `0.2.0` release.

**Pull request merged**